### PR TITLE
Blaze slayer fixes and features

### PR DESCRIPTION
**Pull request by Desco1**
Fixes blaze slayer features being broken and adds 3 new features to blaze slayer.
These are not meant to change how slayer detection or any of the other features work internally to not conflict with the feature/slayer branch. Instead, these are meant to be a temporary solution until SlayerFeatures gets rewritten.

Actual changes:
- Changed timerRegex to properly match all Inferno Demonlord bosses after the demons phase (it also matches demons now, for the Attunement Display feature)
- Updated Slayer Boss Hitbox to also draw around Inferno Demonlord minibosses
- Attunement Display: recolors the Inferno Demonlord or the appropiate demon according to the correct Attunement to use. Unlike the Voidgloom's phase display, the colors are not configurable because I don't think there is a reason since they match the Attunement. Ashen might be too hard to view compared to the rest however
- Ignore pacified blazes: stops rendering blaze entities further from 3 blocks from the player when Smoldering Polarization is active and the Inferno Demonlord is spawned
- Fire pit warning: displays a title and plays a ping when standing in the fire pit that the Inferno Demonlord spawns (that does a billion damage)

Note about the Slayer RNGesus Meter: still broken after the chat message changed. While fixable by keeping track of the open gui, checking if it's a slayer RNG Meter, checking which drop is selected and saving how much is the target XP of the drop, I figured that it'd be easier to save and download the XP required for each drop and use the chat message after changing RNG Meter drop to get it instead.

### CODE REVIEW

File: src/main/kotlin/gg/skytils/skytilsmod/core/Config.kt
Comment (My-Name-Is-Jeff) [inline src/main/kotlin/gg/skytils/skytilsmod/core/Config.kt]: most of the feature names are like capitalized except for words like a, of, etc

File: src/main/kotlin/gg/skytils/skytilsmod/core/Config.kt
Comment (My-Name-Is-Jeff) [inline src/main/kotlin/gg/skytils/skytilsmod/core/Config.kt]: most of the feature names are like capitalized except for words like a, of, etc

File: src/main/kotlin/gg/skytils/skytilsmod/mixins/hooks/renderer/RendererLivingEntityHook.kt
Comment (My-Name-Is-Jeff) [inline src/main/kotlin/gg/skytils/skytilsmod/mixins/hooks/renderer/RendererLivingEntityHook.kt:59]: in this case, could try making alpha higher

**Review by My-Name-Is-Jeff — APPROVED** [review]
Thanks for the pull request! I'm not too familiar with blaze slayer so I'll have Sychic check too.

Logically everything makes sense so

File: src/main/kotlin/gg/skytils/skytilsmod/mixins/hooks/renderer/RendererLivingEntityHook.kt
Comment (Desco1) [reply src/main/kotlin/gg/skytils/skytilsmod/mixins/hooks/renderer/RendererLivingEntityHook.kt:59]: I have tried multiple alpha values for every attunement and different colors for Ashen specifically, but in the end this configuration seems to be the clearer since the alpha on the color multiplier doesn't quite translate to a clearer value. The animation when an entity is hurt uses 30% alpha in vanilla and it seems to be the clearest the color can get, which corresponds to 169 alpha on the hook. For Ashen I have tried the vanilla §7 and §0 but none of them look better than §8 in my opinion.

File: src/main/kotlin/gg/skytils/skytilsmod/core/Config.kt
Comment (Sychic) [inline src/main/kotlin/gg/skytils/skytilsmod/core/Config.kt:2457]: Is this supposed to be "faraway"?

File: src/main/kotlin/gg/skytils/skytilsmod/features/impl/misc/SlayerFeatures.kt
Comment (Sychic) [inline src/main/kotlin/gg/skytils/skytilsmod/features/impl/misc/SlayerFeatures.kt]: Why not put the `substringBefore` with the declaration?

File: src/main/kotlin/gg/skytils/skytilsmod/features/impl/misc/SlayerFeatures.kt
Comment (Sychic) [inline src/main/kotlin/gg/skytils/skytilsmod/features/impl/misc/SlayerFeatures.kt:1227]: How often are these accessed? May be worth handling it somewhere else.

File: src/main/kotlin/gg/skytils/skytilsmod/features/impl/misc/SlayerFeatures.kt
Comment (Sychic) [inline src/main/kotlin/gg/skytils/skytilsmod/features/impl/misc/SlayerFeatures.kt:1232]: Could create a data class if you need but this works.

Comment (Desco1) [reply src/main/kotlin/gg/skytils/skytilsmod/features/impl/misc/SlayerFeatures.kt:1232]: The class alternative was just in case the initial demon + timer detection failed, to instead detect the armor stand after a small delay from the demon detection, but delaying the TickTask from 5 ticks to 10 has worked consistently without any problems. All of those variables do look kinda ugly but I guess doesn't matter much since the demons don't have many mechanics

File: src/main/kotlin/gg/skytils/skytilsmod/features/impl/misc/SlayerFeatures.kt
Comment (Desco1) [reply src/main/kotlin/gg/skytils/skytilsmod/features/impl/misc/SlayerFeatures.kt:1227]: As often as it gets; I haven't had any problems of performance loss but could check for armor stands being renamed from the packet listener to update the relevant entity. Would rather make a few more tests with a profiler but I don't have enough time to make a proper extensive test atm

File: src/main/kotlin/gg/skytils/skytilsmod/core/Config.kt
Comment (Desco1) [reply src/main/kotlin/gg/skytils/skytilsmod/core/Config.kt:2457]: Yes, nearby blazes aren't hidden because the player can still interact with them and also because there might be a bug with minibosses still moving towards the player but not attacking. The range at which the blazes are hidden is just 3 blocks away, but I couldn't find a more appropiate word for that distance

File: src/main/kotlin/gg/skytils/skytilsmod/features/impl/misc/SlayerFeatures.kt
Comment (Desco1) [reply src/main/kotlin/gg/skytils/skytilsmod/features/impl/misc/SlayerFeatures.kt]: Forgot to change that after committing, will be fixed after I rework the relevantEntity and relevantColor getters or I do some profiler tests

File: src/main/kotlin/gg/skytils/skytilsmod/core/Config.kt
Comment (Sychic) [reply src/main/kotlin/gg/skytils/skytilsmod/core/Config.kt:2457]: I mean specifically "faraway" as opposed to "far away". Not trying to nitpick but it seems like a typo maybe?

Comment (Desco1) [reply src/main/kotlin/gg/skytils/skytilsmod/core/Config.kt:2457]: Both of them are correct and mean pretty much the same thing, although "faraway" is an adjective and "far away" is an adverb phrase. Could also be rephrased as "...blazes that are far away..."

File: src/main/kotlin/gg/skytils/skytilsmod/features/impl/misc/SlayerFeatures.kt
Comment (Desco1) [reply src/main/kotlin/gg/skytils/skytilsmod/features/impl/misc/SlayerFeatures.kt:1227]: After around half an hour of each boss being spawned, neither enderman nor blaze bosses made enough of an impact to appear in the profiler. I haven't noticed any increase on CPU or Memory usage or any FPS drop.
Although a single test isn't enough to decide on the impact of something like this, I think it's more worthwhile to manage all potential bottlenecks in a full rewrite of the file in the feature/slayer branch.

File: src/main/kotlin/gg/skytils/skytilsmod/features/impl/misc/SlayerFeatures.kt
Comment (Sychic) [inline src/main/kotlin/gg/skytils/skytilsmod/features/impl/misc/SlayerFeatures.kt]: Does this substring the same way it used to? Looks like one was removed.

Comment (Desco1) [reply src/main/kotlin/gg/skytils/skytilsmod/features/impl/misc/SlayerFeatures.kt]: Yes, it grabs the first word from the boss name and properly gets the expected hp, just without the redundant variable assignment. Compared to main branch, the system is used for the "Use Hits to Detect Slayer" feature, but it now works for all slayers instead of only Voidgloom.

**Review by Sychic — APPROVED** [review]
LGTM, assuming you've tested it

**Comment by Desco1** [comment]
Before merging there are a few things that could be considered:
- RNG Meter display still broken after chat, which could be fixed easiest in my opinion by downloading the required xp for each drop, a few more details in my first message
- Feature: open RNG Meter menu after a boss if the selected boss has dropped the selected item. Can easily be done by comparing stored xp and sending the /cb command of the message, but not sure if it could somehow be considered a cheat which is why I haven't added it yet.
- Feature: a visual display for every blaze that is affected by the 2x damage from the Moody Grappleshot. Not too useful unless abusing tactical insertion, but might be neat to have? From a bit of testing though, it doesn't seem that there is any easy way to detect when a blaze is being pulled
I don't think either of those are critical enough so merging and then following this pr with another in some future is also fine

**Comment by Sychic** [comment]
>     * RNG Meter display still broken after chat, which could be fixed easiest in my opinion by downloading the required xp for each drop, a few more details in my first message

How did the message change? I haven't done slayer in forever so I have no clue what it is now or when it changed. Did it change from percentage with a bar?

>     * Feature: open RNG Meter menu after a boss if the selected boss has dropped the selected item. Can easily be done by comparing stored xp and sending the /cb command of the message, but not sure if it could somehow be considered a cheat which is why I haven't added it yet.

We could do something similar to the "open maddox batphone" feature where you just click anywhere instead of clicking on a specific bit.

>     * Feature: a visual display for every blaze that is affected by the 2x damage from the Moody Grappleshot. Not too useful unless abusing tactical insertion, but might be neat to have? From a bit of testing though, it doesn't seem that there is any easy way to detect when a blaze is being pulled

We could do something where the blaze itself checks if there's the anvil + the lead nearby (there's got to be some invisible entity that the lead is tied to, no clue what it could be tho).

>       I don't think either of those are critical enough so merging and then following this pr with another in some future is also fine

Also works for me.

**Comment by Desco1** [comment]
> RNG meter chat message

[Reference](https://i.imgur.com/nlXiwds.png). Yes, it now only displays the stored XP, not which drop was selected nor % to it or the next. When the RNG Meter is complete, the drop is shown in chat without magic find, but if it drops by itself, even when selected, the magic find will be shown.
[More reference](https://i.imgur.com/MXb7I4M.png). There is also this chat message shown when a drop is selected from the meter gui, which also shows the slayer it is from. All drops have the same XP for every single player, which is why I think it could be downloaded from server at startup.
Another alternative could be permanently looking for the RNG Meter gui and grabbing the required XP from [there](https://i.imgur.com/k5dlahU.png), but it gets [truncated when actually selected](https://i.imgur.com/PI3AIcs.png), which is pretty annoying
> RNG meter menu like maddox phone

Thought of that, but I didn't want to deal with the consequences of having both messages at once. The Stored XP message could not show when autoslayer is disabled, although I may have just missed it. Could also make it so a different mouse button is a different menu?
> Hooks

When a player uses the hook an S1BPacketEntityAttach gets sent, `leash` is not 0, `vehicleEntityId` is the player who used it and `entityID` is a bat, which most of the times is not in the world yet and its spawn has to be handled in a different event. I've tried recreating vanilla collissions from that bat to blazes but I didn't get any.
I have tried looking for the "snap" effect where some blazes teleport slightly to the hook and checking if blazes have the same velocity as the bat (they do in X and Z, but it is always 0); but I didn't get consistent results for that detection.
Also, the hook 2x damage doesn't last forever, but that can be easily tested compared to the detection

**Comment by Sychic** [comment]
> RNG meter chat message

We should be able to get all the data when the player is actually clicking (though that would take a bit of trust in hypixel to actually receive the click). It's probably just more reliable to download from the data repo and look for when the item is updated to tell if hypixel actually received the click to update the RNG item. 
Though this should probably be separate from all these changes.

> RNG meter menu like maddox phone

Wdym both messages at once? I don't think there's a specific scenario when you would have both the maddox menu and the rng menu (since the maddox one should be entirely manual from using the batphone). We would just keep the most recent `/cb` usage and execute that.

> Hooks

We could use the bat itself to look for blazes near itself (using `Entity#getEntitiesWithinAABBExcludingEntity`) and set a flag in them to tell somewhere else to render them differently. Not entirely version agnostic but not really any better way to do it (though adapting mixins should be pretty simple)

**Comment by Desco1** [comment]
> RNG menu like maddox phone

There isn't a situation where both messages would be active at the same time but I figured it's better to save players from themselves.

In any case, I agree that this PR should not be expanded with more features and instead I'll follow up with the rest of them in some future. Should be ready to merge

**Pull request merged**
